File: plugins/probe-rat-layer/probe_layer/core/graph_coloring.py
"""DSATUR 着色（方法 A：对比基线，支持受限着色）。"""
from __future__ import annotations

from ..model import ConflictGraph


class UncolorableError(RuntimeError):
    pass


def dsatur_color(graph: ConflictGraph,
                 max_layers: int | None = None,
                 allowed: dict[str, set[int]] | None = None,
                 initial: dict[str, int] | None = None) -> dict[str, int]:
    """DSATUR 变体。

    allowed: wire_id -> 允许的层集合（受限着色）；None 表示任意层。
    initial: 预着色（增量修复时冻结的线）。
    """
    colors: dict[str, int] = dict(initial or {})
    uncolored = set(graph.nodes()) - set(colors.keys())

    while uncolored:
        best: str | None = None
        best_sat = -1
        best_deg = -1
        for n in uncolored:
            sat = len({colors[m] for m in graph.neighbors(n) if m in colors})
            deg = len(graph.neighbors(n))
            if sat > best_sat or (sat == best_sat and deg > best_deg):
                best_sat, best_deg, best = sat, deg, n

        used = {colors[m] for m in graph.neighbors(best) if m in colors}
        domain = allowed.get(best) if allowed else None
        if domain is not None:
            avail = sorted(domain - used)
            if not avail:
                raise UncolorableError(f"节点 {best} 无可用层（允许 {sorted(domain)}，邻用 {sorted(used)}）")
            colors[best] = avail[0]
        else:
            c = 0
            while c in used:
                c += 1
            colors[best] = c
        uncolored.remove(best)

    if max_layers is not None:
        used_max = max(colors.values(), default=0) + 1
        if used_max > max_layers:
            raise UncolorableError(f"着色需要 {used_max} 层，超过 max_layers={max_layers}")
    return colors


def minimize_layers(assignment: dict[str, int]) -> int:
    return (max(assignment.values()) + 1) if assignment else 0

File: core-plugins/probe-rat-layer/src/graph_coloring.rs
//! DSATUR 着色（方法 A：对比基线，支持受限着色；移植自 Python `core/graph_coloring.py`）。

use crate::model::ConflictGraph;
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone)]
pub struct UncolorableError(pub String);

impl std::fmt::Display for UncolorableError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for UncolorableError {}

/// DSATUR 变体。`allowed`（wire_id -> 允许层集合）None 表示任意层；`initial` 为预着色。
pub fn dsatur_color(
    graph: &ConflictGraph,
    max_layers: Option<i64>,
    allowed: Option<&HashMap<String, HashSet<i64>>>,
    initial: Option<&HashMap<String, i64>>,
) -> Result<HashMap<String, i64>, UncolorableError> {
    let mut colors: HashMap<String, i64> = initial.cloned().unwrap_or_default();
    let mut uncolored: Vec<String> = graph
        .nodes()
        .into_iter()
        .filter(|n| !colors.contains_key(n))
        .collect();

    while !uncolored.is_empty() {
        // 选饱和度最大、其次度数最大的未着色节点
        let mut best: Option<String> = None;
        let mut best_sat = -1;
        let mut best_deg = -1;
        for n in &uncolored {
            let mut seen: HashSet<i64> = HashSet::new();
            for m in graph.neighbors(n) {
                if let Some(&c) = colors.get(&m) {
                    seen.insert(c);
                }
            }
            let sat = seen.len() as i64;
            let deg = graph.neighbors(n).len() as i64;
            if sat > best_sat || (sat == best_sat && deg > best_deg) {
                best_sat = sat;
                best_deg = deg;
                best = Some(n.clone());
            }
        }
        let best_node = best.ok_or_else(|| UncolorableError("无节点可着色".into()))?;

        let mut used: HashSet<i64> = HashSet::new();
        for m in graph.neighbors(&best_node) {
            if let Some(&c) = colors.get(&m) {
                used.insert(c);
            }
        }
        let domain = allowed.and_then(|a| a.get(&best_node));
        let color: i64 = if let Some(domain) = domain {
            let mut avail: Vec<i64> = domain.iter().copied().filter(|c| !used.contains(c)).collect();
            avail.sort();
            if avail.is_empty() {
                let mut d: Vec<i64> = domain.iter().copied().collect();
                d.sort();
                let mut u: Vec<i64> = used.iter().copied().collect();
                u.sort();
                return Err(UncolorableError(format!(
                    "节点 {} 无可用层（允许 {:?}，邻用 {:?}）",
                    best_node, d, u
                )));
            }
            avail[0]
        } else {
            let mut c = 0;
            while used.contains(&c) {
                c += 1;
            }
            c
        };
        colors.insert(best_node.clone(), color);
        uncolored.retain(|n| n != &best_node);
    }

    if let Some(max_layers) = max_layers {
        let used_max = colors.values().copied().max().unwrap_or(0) + 1;
        if used_max > max_layers {
            return Err(UncolorableError(format!(
                "着色需要 {} 层，超过 max_layers={max_layers}",
                used_max
            )));
        }
    }
    Ok(colors)
}

pub fn minimize_layers(assignment: &HashMap<String, i64>) -> i64 {
    assignment.values().copied().max().map(|m| m + 1).unwrap_or(0)
}
